cs3api4lab/api/cs3_share_api.py

```diff
@@ -0,0 +1,364 @@
+import cs3.storage.provider.v1beta1.provider_api_pb2 as storage_provider
+import cs3.storage.provider.v1beta1.resources_pb2 as storage_resources
+import cs3.identity.user.v1beta1.resources_pb2 as identity_res
+import cs3.rpc.v1beta1.code_pb2 as cs3_code
+
+from IPython.utils import tz
+
+from cs3api4lab.auth.authenticator import Authenticator
+from cs3api4lab.api.cs3_file_api import Cs3FileApi
+from cs3api4lab.api.file_utils import FileUtils
+from cs3api4lab.common.strings import *
+from cs3api4lab.config.config_manager import Cs3ConfigManager
+import cs3.gateway.v1beta1.gateway_api_pb2_grpc as grpc_gateway
+from cs3api4lab.auth.channel_connector import ChannelConnector
+
+
+class Cs3ShareApi:
+    gateway_stub = None
+    log = None
+    auth = None
+    config = {}
+    file_api = None
+    date_fmt = '%d-%b-%Y (%H:%M:%S.%f)'
+
+    def __init__(self, log):
+        self.config = Cs3ConfigManager().config
+        self.auth = Authenticator()
+        self.file_api = Cs3FileApi(log)
+        self.gateway_stub = grpc_gateway.GatewayAPIStub(ChannelConnector().channel)
+        self.log = log
+        return
+
+    def create(self, endpoint, file_path, grantee, idp, role=Role.VIEWER, grantee_type=Grantee.USER):
+        if grantee is None:
+            raise Exception("Grantee was not provided")
+        share_permissions = self._get_share_permissions(role)
+        grantee_type_enum = self._get_grantee_type(grantee_type)
+        share_grant = self._get_share_grant(grantee_type_enum, share_permissions, idp, grantee)
+        resource_info = self._get_resource_info(endpoint, file_path)
+        share_request = sharing.CreateShareRequest(resource_info=resource_info, grant=share_grant)
+        token = self.get_token()
+        share_response = self.gateway_stub.CreateShare(request=share_request,
+                                                       metadata=[('x-access-token', token)])
+        self._check_response_code(share_response)
+        self.log.info("Share created")
+        self.log.info(share_response)
+        return self._map_given_share(share_response.share)
+
+    def list(self):
+        list_response = self._list()
+        return self._map_given_shares(list_response)
+
+    def list_dir_model(self):
+        list_response = self._list()
+        return self._map_shares_to_dir_model(list_response)
+
+    def _list(self):
+        # todo filters
+        list_request = sharing.ListSharesRequest()
+        list_response = self.gateway_stub.ListShares(request=list_request,
+                                                     metadata=[('x-access-token', self.get_token())])
+        self._check_response_code(list_response)
+        self.log.info("List shares response for user: " + self.config['client_id'])
+        self.log.info(list_request)
+        return list_response
+
+    def list_grantees_for_file(self, file_path):
+        stat = self.file_api.stat(file_path, self.config['client_id'])
```

Comment:
There's no way to filter by path? Do we really need to have the resource id?
If that is the case, then the API should expose resource IDs and not paths, as this seems an unnecessary round trip for information that we should already have.
Ofc that dealing with paths makes more sense, though...

Reply:
Initially all of the shares were fetched and filtered manually by file path. With your advice I used a built-in CS3 Api filter, which made more sense. However to retrieve a share using a filter, the resource ID including storage ID and file ID is needed. Unfortunatelly it turned out, that the frontend does not have the knowledge about storage IDs. The request containes only file path as it was before, but the 'round-trip' is necessary to make it work if we want to use the filter

Reply:
Ok, let's keep it like this for now. I would say that the front end needs to know the ID, so this will have to be added. (Or we should ask for the CS3APIs to support filter by path.
Put a `FIXME` in the code so that we show that we know this is not great.